### PR TITLE
bluetooth: mesh: align sensor/column reporting

### PR DESCRIPTION
**Pull request by ludvigsj**
This aligns sensor and column reporting for sensors with any number of
channels.

Signed-off-by: Ludvig Samuelsen Jordet <ludvig.jordet@nordicsemi.no>

test-sdk-nrf: sdk-nrf-PR-7676

### CODE REVIEW

**Comment by Andrewpini** [comment]
Might be me not seeing the full context, but does not the `bt_mesh_sensor_cli_series_entries_get` in sensor_cli.c require some refactoring after the spec change as well?

**Comment by ludvigsj** [comment]
> Might be me not seeing the full context, but does not the `bt_mesh_sensor_cli_series_entries_get` in sensor_cli.c require some refactoring after the spec change as well?

Oh wow, you are of course correct. Will fix.

File: subsys/bluetooth/mesh/sensor_srv.c [hunk 1]
```diff
@@ -310,12 +333,44 @@ static int handle_series_get(struct bt_mesh_model *model, struct bt_mesh_msg_ctx
...
+	if (sensor->type->channel_count < 3) {
+		uint16_t start = 0;
+		uint16_t end = sensor->series.column_count - 1;
+
+		if (buf->len == sizeof(uint16_t) * 2) {
+			start = net_buf_simple_pull_le16(buf);
+			end = net_buf_simple_pull_le16(buf);
+			if (start >= sensor->series.column_count ||
+			    end >= sensor->series.column_count ||
+			    start < end) {
```
Comment (PavelVPV) [inline subsys/bluetooth/mesh/sensor_srv.c:350]:
```suggestion
			    end < start) {
```
?

File: subsys/bluetooth/mesh/sensor_types.c [hunk 2]
```diff
@@ -769,7 +769,6 @@ SENSOR_TYPE(time_since_presence_detected) = {
  ******************************************************************************/
 SENSOR_TYPE(avg_amb_temp_in_day) = {
 	.id = BT_MESH_PROP_ID_AVG_AMB_TEMP_IN_A_PERIOD_OF_DAY,
-	.flags = BT_MESH_SENSOR_TYPE_FLAG_SERIES,
```
Comment (PavelVPV) [inline subsys/bluetooth/mesh/sensor_types.c:772]: Do we need to keep `flags` field and `BT_MESH_SENSOR_TYPE_FLAG_SERIES`?

**Review by PavelVPV — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/mesh/sensor_srv.c (hunk 1 above)
Comment (Andrewpini) [reply subsys/bluetooth/mesh/sensor_srv.c:350]: Got me a little dizzy aswell. Is it related to this?
image

Comment (PavelVPV) [reply subsys/bluetooth/mesh/sensor_srv.c:350]: If I don't miss anything, the for-loop below expects `start` to be before `end`.

File: subsys/bluetooth/mesh/sensor_srv.c [hunk 3]
```diff
@@ -310,12 +333,44 @@ static int handle_series_get(struct bt_mesh_model *model, struct bt_mesh_msg_ctx
...
+
+		for (uint32_t i = start; i <= end; i++) {
+			int err = sensor_column_value_encode(&rsp, srv, sensor,
+							     ctx, i);
+
+			if (err) {
+				BT_WARN("Column encode failed");
+				return err;
+			}
+		}
```
Comment (Andrewpini) [inline subsys/bluetooth/mesh/sensor_srv.c:368]: 4.3.1.2.8 from spec rev:
image
Does this implementation cover this? Will a get message without specifying start/end boundaries ensure that a status response contains N number of columns that fits within a payload if the total amount of available columns are greater than what fits into a payload?

File: subsys/bluetooth/mesh/sensor_srv.c (hunk 1 above)
Comment (Andrewpini) [reply subsys/bluetooth/mesh/sensor_srv.c:350]: Then comes my follow-up Q: Is idx 0 equal to the most recent column?

File: subsys/bluetooth/mesh/sensor_cli.c [hunk 4]
```diff
@@ -189,14 +189,20 @@ static int parse_series_entry(const struct bt_mesh_sensor_type *type,
 {
 	int err;
 
-	err = sensor_ch_decode(buf, col_format, &entry->column.start);
+	if (col_format == NULL) {
+		// Indexed column, decode only value
+		entry->column = NULL;
```
[the remark this answers is not in the document]
Comment (ludvigsj) [reply subsys/bluetooth/mesh/sensor_cli.c:194]: The idea was that it is confusing if a `column` is passed to the callback when there really isn't a defined column (for indexed columns)

File: subsys/bluetooth/mesh/sensor_types.c (hunk 2 above)
Comment (ludvigsj) [reply subsys/bluetooth/mesh/sensor_types.c:772]: As far as I could tell, they're not used anywhere?

File: subsys/bluetooth/mesh/sensor_cli.c [hunk 5]
```diff
@@ -258,8 +263,9 @@ static int handle_column_status(struct bt_mesh_model *model, struct bt_mesh_msg_
...
 	if (bt_mesh_msg_ack_ctx_match(&cli->ack_ctx, BT_MESH_SENSOR_OP_COLUMN_STATUS, ctx->addr,
 				      (void **)&rsp) &&
-	    rsp->col->start.val1 == entry.column.start.val1 &&
-	    rsp->col->start.val2 == entry.column.start.val2) {
+	    ((entry.column == NULL) ||
+	    (rsp->col->start.val1 == entry.column->start.val1 &&
+	     rsp->col->start.val2 == entry.column->start.val2))) {
 		if (err) {
 			rsp->count = 0;
 		} else {
```
[the remark this answers is not in the document]
Comment (ludvigsj) [reply subsys/bluetooth/mesh/sensor_cli.c:271]: Right. I think I'll just move the column back inside the entry struct, seems too much hassle to pull it out like this.

File: doc/nrf/libraries/bluetooth_services/mesh/sensor.rst [hunk 6]
```diff
@@ -253,15 +254,16 @@ Publication may also be forced by calling :c:func:`bt_mesh_sensor_srv_sample`, w
 Sensor series
 *************
 
-Sensor series data is organized into a static set of columns, specified at init.
-The sensor series :c:member:`bt_mesh_sensor_series.get` callback must be implemented to enable the sensor's series data feature.
-Only some sensor types support series access, see the sensor type's documentation.
+Sensor series data may be provided for all sensor types.
+:c:member:`bt_mesh_sensor_series.column_count` must be specified and the sensor series :c:member:`bt_mesh_sensor_series.get` callback must be implemented to enable the sensor's series data feature.
+
+For sensor types with more than two channels, the series data is organized into a static set of colums, specified at init.
```
Comment (mia-ko) [inline doc/nrf/libraries/bluetooth_services/mesh/sensor.rst:260]:
```suggestion
Sensor series data can be provided for all sensor types.
To enable the sensor's series data feature, :c:member:`bt_mesh_sensor_series.column_count` must be specified and the sensor series :c:member:`bt_mesh_sensor_series.get` callback must be implemented.

For sensor types with more than two channels, the series data is organized into a static set of columns specified at init.
```

File: doc/nrf/libraries/bluetooth_services/mesh/sensor.rst [hunk 7]
```diff
@@ -253,15 +254,16 @@ Publication may also be forced by calling :c:func:`bt_mesh_sensor_srv_sample`, w
...
 The format of the column may be queried with :c:func:`bt_mesh_sensor_column_format_get`.
 
-The ``get`` callback gets called with a direct pointer to one of the columns in the column list, and is expected to fill the ``value`` parameter with sensor data for the specified column.
+The ``get`` callback gets called with an index of one of the columns, and is expected to fill the ``value`` parameter with sensor data for the specified column.
 If a Sensor Client requests a series of columns, the callback may be called repeatedly, requesting data from each column.
 
-Example: Average ambient temperature in a period of day as a sensor series:
+Example: Three channel sensor (Average ambient temperature in a period of day) as a sensor series:
```
Comment (mia-ko) [inline doc/nrf/libraries/bluetooth_services/mesh/sensor.rst:266]:
```suggestion
Example: A three-channel sensor (average ambient temperature in a period of day) as a sensor series:
```

File: doc/nrf/libraries/bluetooth_services/mesh/sensor.rst [hunk 8]
```diff
@@ -286,17 +288,37 @@ Example: Average ambient temperature in a period of day as a sensor series:
...
+                     uint32_t column_index, struct sensor_value *value)
+   {
+       value[0] = avg_temp[column_index];
+       value[1] = columns[column_index].start;
+       value[2] = columns[column_index].end;
+
+       return 0;
+   }
+
+Example: Single-channel sensor (Motion sensed) as a sensor series:
```
Comment (mia-ko) [inline doc/nrf/libraries/bluetooth_services/mesh/sensor.rst:300]:
```suggestion
Example: Single-channel sensor (motion sensed) as a sensor series:
```

**Review by mia-ko — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/mesh/sensor_srv.c [hunk 9]
```diff
@@ -252,7 +252,34 @@ static int handle_column_get(struct bt_mesh_model *model, struct bt_mesh_msg_ctx
...
+	}
+
+	if (sensor->type->channel_count < 3) {
+		/* For sensors with one or two channels, pull the column index
+		 * directly from the message
+		 */
+		uint32_t col_index = net_buf_simple_pull_le16(buf);
+
+		if (col_index >= sensor->series.column_count) {
+			BT_WARN("Column out of range in 0x%04x",
```
[the remark this answers is not in the document]
Comment (PavelVPV) [reply subsys/bluetooth/mesh/sensor_srv.c:267]: Not addressed, but up to you.

File: subsys/bluetooth/mesh/sensor_srv.c [hunk 10]
```diff
@@ -310,12 +333,44 @@ static int handle_series_get(struct bt_mesh_model *model, struct bt_mesh_msg_ctx
...
+			BT_WARN("Invalid length (%u)", buf->len);
+			return -EMSGSIZE;
+		}
+
+		for (uint32_t i = start; i <= end; i++) {
+			int err = sensor_column_value_encode(&rsp, srv, sensor,
+							     ctx, i);
+
+			if (err) {
+				BT_WARN("Column encode failed");
```
[the remark this answers is not in the document]
Comment (PavelVPV) [reply subsys/bluetooth/mesh/sensor_srv.c:384]: Not addressed, but up to you.

**Review by PavelVPV — APPROVED** [review]

File: subsys/bluetooth/mesh/sensor_types.c (hunk 2 above)
Comment (ludvigsj) [reply subsys/bluetooth/mesh/sensor_types.c:772]: We're changing the API here either way and need to mention that in the release notes (series CB now receiving index insted of column pointer).

**Review by mia-ko — COMMENTED** [review]
Btw, are you adding an entry to the changelog in this PR?

**Pull request merged**